src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3941,38 +3962,33 @@ public void EnsureInitialized()
         /// </summary>
         [DebuggerStepThrough]
         [DebuggerHidden]
-        internal object? CreateInstanceDefaultCtor(bool publicOnly, bool skipCheckThis, bool fillCache, bool wrapExceptions)
+        internal unsafe object? CreateInstanceDefaultCtor(bool publicOnly, bool skipCheckThis, bool fillCache, bool wrapExceptions)
         {
-            // Call the cached
-            if (GenericCache is ActivatorCache cacheEntry)
+            // Call the cached factory if it exists
+            if (GenericCache is ActivatorCache cache)
             {
-                cacheEntry.EnsureInitialized();
+                cache.EnsureInitialized(this);
+
+                Debug.Assert(cache._pfnNewobj != null);
+                Debug.Assert(cache._pfnCtor != null);
 
-                if (publicOnly)
+                if (!cache._ctorIsPublic && publicOnly)
                 {
-                    if (cacheEntry._ctor != null &&
-                        (cacheEntry._ctorAttributes & MethodAttributes.MemberAccessMask) != MethodAttributes.Public)
-                    {
-                        throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, this));
-                    }
+                    throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, this));
                 }
 
-                // Allocate empty object and call the default constructor if present.
-                object instance = RuntimeTypeHandle.Allocate(this);
-                Debug.Assert(cacheEntry._ctor != null || IsValueType);
-                if (cacheEntry._ctor != null)
+                try
                 {
-                    try
-                    {
-                        cacheEntry._ctor(instance);
-                    }
-                    catch (Exception e) when (wrapExceptions)
-                    {
-                        throw new TargetInvocationException(e);
-                    }
-                }
+                    object? obj = cache._pfnNewobj(cache._pMT);
```

Comment:
The existing implementation has the object allocation outside the `TargetInvocationException` wrapping